lib/Dialect/Arc/Transforms/GroupResetsAndEnables.cpp

```diff
@@ -0,0 +1,160 @@
+    ClockTreeOp clockTreeOp = dyn_cast<ClockTreeOp>(*op);
+
+    // Group similar resets into single IfOps
+    // Create a list of reset values and map from them to the states they reset
+    llvm::MapVector<mlir::Value, SmallVector<scf::IfOp>> resetMap;
+    auto ifOps = clockTreeOp.getBody().getOps<scf::IfOp>();
+
+    for (auto ifOp : ifOps)
+      resetMap[ifOp.getCondition()].push_back(ifOp);
+
+    // Combine IfOps
+    bool changed = false;
+    for (auto [cond, oldOps] : resetMap) {
+      if (oldOps.size() > 1) {
+        auto iteratorStart = oldOps.rbegin();
+        scf::IfOp lastIfOp = *(iteratorStart++);
+        for (auto thisOp = iteratorStart; thisOp != oldOps.rend(); thisOp++) {
+          // Inline the before and after region inside the original If
+          rewriter.eraseOp(thisOp->thenBlock()->getTerminator());
+          rewriter.inlineBlockBefore(thisOp->thenBlock(),
+                                     &lastIfOp.thenBlock()->front());
+          // Check we're not inlining an empty block
+          if (!thisOp->elseBlock()->empty()) {
+            rewriter.eraseOp(thisOp->elseBlock()->getTerminator());
+            rewriter.inlineBlockBefore(thisOp->elseBlock(),
+                                       &lastIfOp.elseBlock()->front());
+          }
+          rewriter.eraseOp(*thisOp);
+          changed = true;
+        }
+      }
+    }
+    if (!changed)
+      return failure();
+    return success();
+  }
+};
+
+class EnableGroupingPattern : public RewritePattern {
+public:
+  using RewritePattern::RewritePattern;
+  EnableGroupingPattern(PatternBenefit benefit, MLIRContext *context)
+      : RewritePattern(ClockTreeOp::getOperationName(), benefit, context) {}
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    ClockTreeOp clockTreeOp = dyn_cast<ClockTreeOp>(*op);
+
+    // Amass regions that we want to group enables in
+    SmallVector<Region *> groupingRegions;
+    groupingRegions.push_back(&(clockTreeOp.getBody()));
+    for (auto ifOp : clockTreeOp.getBody().getOps<scf::IfOp>()) {
+      groupingRegions.push_back(&ifOp.getThenRegion());
+      groupingRegions.push_back(&ifOp.getElseRegion());
+    }
+
+    bool changed = false;
+    for (auto *region : groupingRegions) {
+      llvm::MapVector<mlir::Value, SmallVector<StateWriteOp>> enableMap;
+      auto writeOps = region->getOps<StateWriteOp>();
+      for (auto writeOp : writeOps) {
+        if (writeOp.getCondition())
+          enableMap[writeOp.getCondition()].push_back(writeOp);
+      }
+      for (auto [enable, writeOps] : enableMap) {
+        // Only group if multiple writes share an enable
+        if (writeOps.size() > 1) {
+          if (isa<scf::IfOp>(region->getParentOp())) {
+            rewriter.setInsertionPoint(region->back().getTerminator());
+          } else {
+            rewriter.setInsertionPointToEnd(&region->back());
+          }
```

Comment:
I wanted to suggest `rewriter.setInsertionPointAfter(region->back().without_terminator().end())`, but it seems like it wasn't updated to handle the terminator trait correctly, i.e., does the wrong thing for blocks that don't have a terminator. Nice that you already considered this here.
Nit: since there is only one statement in each branch, you could remove the curly braces.

Reply:
Yeah, the unusual block behaviour without terminators was a bit of a head-scratcher!